crypto/aes/asm/aes-riscv32-zkn.pl

```diff
@@ -0,0 +1,1061 @@
+    }
+    my $regstr = $reglookup{$reg};
+    if (!($regstr =~ /^x([0-9]+)$/)) {
+        die("Could not process register ".$reg);
+    }
+    return $1;
+}
+
+sub aes32dsi {
+    # Encoding for aes32dsi rd, rs1, rs2, bs instruction on RV32
+    #                bs_XXXXX_ rs2 _ rs1 _XXX_ rd  _XXXXXXX
+    my $template = 0b00_10101_00000_00000_000_00000_0110011;
+    my $rd = read_reg shift;
+    my $rs1 = read_reg shift;
+    my $rs2 = read_reg shift;
+    my $bs = shift;
+
+    return ".word ".($template | ($bs << 30) | ($rs2 << 20) | ($rs1 << 15) | ($rd << 7));
+}
+
+sub aes32dsmi {
+    # Encoding for aes32dsmi rd, rs1, rs2, bs instruction on RV32
+    #                bs_XXXXX_ rs2 _ rs1 _XXX_ rd  _XXXXXXX
+    my $template = 0b00_10111_00000_00000_000_00000_0110011;
+    my $rd = read_reg shift;
+    my $rs1 = read_reg shift;
+    my $rs2 = read_reg shift;
+    my $bs = shift;
+
+    return ".word ".($template | ($bs << 30) | ($rs2 << 20) | ($rs1 << 15) | ($rd << 7));
+}
+
+sub aes32esi {
+    # Encoding for aes32esi rd, rs1, rs2, bs instruction on RV32
+    #                bs_XXXXX_ rs2 _ rs1 _XXX_ rd  _XXXXXXX
+    my $template = 0b00_10001_00000_00000_000_00000_0110011;
+    my $rd = read_reg shift;
+    my $rs1 = read_reg shift;
+    my $rs2 = read_reg shift;
+    my $bs = shift;
+
+    return ".word ".($template | ($bs << 30) | ($rs2 << 20) | ($rs1 << 15) | ($rd << 7));
+}
+
+sub aes32esmi {
+    # Encoding for aes32esmi rd, rs1, rs2, bs instruction on RV32
+    #                bs_XXXXX_ rs2 _ rs1 _XXX_ rd  _XXXXXXX
+    my $template = 0b00_10011_00000_00000_000_00000_0110011;
+    my $rd = read_reg shift;
+    my $rs1 = read_reg shift;
+    my $rs2 = read_reg shift;
+    my $bs = shift;
+
+    return ".word ".($template | ($bs << 30) | ($rs2 << 20) | ($rs1 << 15) | ($rd << 7));
+}
+
+sub rori {
+    # Encoding for ror rd, rs1, imm instruction on RV64
+    #                XXXXXXX_shamt_ rs1 _XXX_ rd  _XXXXXXX
+    my $template = 0b0110000_00000_00000_101_00000_0010011;
+    my $rd = read_reg shift;
+    my $rs1 = read_reg shift;
+    my $shamt = shift;
+
+    return ".word ".($template | ($shamt << 20) | ($rs1 << 15) | ($rd << 7));
+}
+
+################################################################################
+# Register assignment for rv32i_zkne_encrypt and rv32i_zknd_decrypt
+################################################################################
+
+# Registers initially to hold AES state (called s0-s3 or y0-y3 elsewhere)
+my ($Q0,$Q1,$Q2,$Q3) = use_regs(6..9);
+
+# Function arguments (x10-x12 are a0-a2 in the ABI)
+# Input block pointer, output block pointer, key pointer
+my ($INP,$OUTP,$KEYP) = use_regs(10..12);
+
+# Registers initially to hold Key
+my ($T0,$T1,$T2,$T3) = use_regs(13..16);
+
+# Loop counter
+my ($loopcntr) = use_regs(30);
+
+################################################################################
+# Utility for rv32i_zksed_encrypt and rv32i_zksed_decrypt
```

Comment:
It looks strange that there are zksed stuff in AES, which is zkn (n is for NIST) but not zks (s is for ShangMi)

Reply:
That was accidentally copied from https://github.com/openssl/openssl/pull/18285. I'll fix it soon.

Reply:
?

Reply:
> I am quite busy recently.

Sorry about this. I will finish my deadline in 3-5 days so I'll fix it and other commends by the end of this month.

Reply:
Fixed now.